pkg/buildermgr/envwatcher.go

```diff
@@ -149,178 +128,126 @@ func (envw *environmentWatcher) getLabels(envName string, envNamespace string, e
-	rv := ""
-	for {
-		wi, err := envw.fissionClient.CoreV1().Environments(metav1.NamespaceAll).Watch(ctx,
-			metav1.ListOptions{
-				ResourceVersion: rv,
-			})
-		if err != nil {
-			if utils.IsNetworkError(err) {
-				envw.logger.Error("encountered network error, retrying later", zap.Error(err))
-				time.Sleep(5 * time.Second)
-				continue
-			}
-			envw.logger.Fatal("error watching environment list", zap.Error(err))
-		}
+func (envw *environmentWatcher) Run(ctx context.Context) {
+	for _, informer := range envw.envWatchInformer {
+		go informer.Run(ctx.Done())
+	}
+}
 
-		for {
-			ev, more := <-wi.ResultChan()
-			if !more {
-				// restart watch from last rv
-				break
-			}
-			if ev.Type == watch.Error {
-				// restart watch from the start
-				rv = ""
-				time.Sleep(time.Second)
-				break
-			}
-			env := ev.Object.(*fv1.Environment)
-			rv = env.ObjectMeta.ResourceVersion
-			envw.sync(ctx)
-		}
+func (envw *environmentWatcher) EnvWatchEventHandlers(ctx context.Context) {
+	for _, informer := range envw.envWatchInformer {
+		informer.AddEventHandler(k8sCache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				envObj := obj.(*fv1.Environment)
+				envw.AddUpdateBuilder(ctx, envObj)
+			},
+			UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+				oldEnvObj := oldObj.(*fv1.Environment)
+				newEnvObj := newObj.(*fv1.Environment)
+				if oldEnvObj.ObjectMeta.ResourceVersion != newEnvObj.ObjectMeta.ResourceVersion {
+					envw.AddUpdateBuilder(ctx, newEnvObj)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				envObj := obj.(*fv1.Environment)
+				envw.DeleteBuilder(ctx, envObj)
+			},
+		})
 	}
 }
 
-func (envw *environmentWatcher) sync(ctx context.Context) {
-	maxRetries := 10
-	for i := 0; i < maxRetries; i++ {
-		envList, err := envw.fissionClient.CoreV1().Environments(metav1.NamespaceAll).List(ctx, metav1.ListOptions{})
-		if err != nil {
-			if utils.IsNetworkError(err) {
-				envw.logger.Error("error syncing environment CRD resources due to network error, retrying later", zap.Error(err))
-				time.Sleep(50 * time.Duration(2*i) * time.Millisecond)
-				continue
+func (envw *environmentWatcher) AddUpdateBuilder(ctx context.Context, env *fv1.Environment) {
+	//builder is not supported with v1 interface and ignore env without builder image
+	if env.Spec.Version != 1 && len(env.Spec.Builder.Image) != 0 {
+		if _, ok := envw.cache[crd.CacheKeyUID(&env.ObjectMeta)]; !ok {
+			builderInfo, err := envw.createBuilder(ctx, env, envw.getNamespace(env))
+			if err != nil {
+				envw.logger.Error("error creating builder service", zap.Error(err))
+				return
 			}
-			envw.logger.Fatal("error syncing environment CRD resources", zap.Error(err))
-		}
-
-		// Create environment builders for all environments
-		for i := range envList.Items {
-			env := envList.Items[i]
+			envw.cache[crd.CacheKeyUID(&env.ObjectMeta)] = builderInfo
+		} else {
+			envw.DeleteBuilderService(ctx, env)
```

Comment:
Can we remove line 170-172 and just call ? 

```
envw.DeleteBuilder(ctx, env) 
```

Rest looks looks fine to me.

Reply:
done